[BUGFIX] Allow "Content-Type" in content

JakeQZ opened the pull request:
Use a regular expression to determine whether the provided HTML contains a
`Content-Type` `meta` element instead of just testing for the string
"Content-Type".  Also allow for case insensitivity.

A valid `Content-Type` `meta` element which is invalidly placed (i.e. within the
`body`) will still be determined to be valid.  This will be addressed separately
for #923.

Add the package `rawr/cross-data-providers` to the development dependencies.
This is useful to generate a matrix of combinations for test data providers.
There are a few existing instances where this could be used to replace code
effectively doing that, but which are not addressed in this commit.

Rename affected data provider methods from `xyzDataProvider` to `provideXyx`,
which is more succinct and a more commonly used nomenclature on other projects.

Fixes #957.

oliverklee commented on src/HtmlProcessor/AbstractHtmlProcessor.php:
```diff
@@ -271,7 +271,8 @@ private function normalizeDocumentType(string $html): string
     {
-        $hasContentTypeMetaTag = \stripos($html, 'Content-Type') !== false;
+        $hasContentTypeMetaTag
+            = \preg_match('%<meta(?=\\s)[^>]*\\shttp-equiv=(["\']?+)Content-Type\\g{-1}[\\s/>]%i', $html) === 1;
```
With the `=== 1`,  we've switched the meaning from "at least one content-type meta tag" to "exactly one content-type meta tag".

Even if multiple occurrences do not make much sense, I'd still prefer us to recognize this as "there is one" by keeping using `!== false`, and also cover this edge case with a test.

oliverklee replied:
I don't quite understand the `(?=\\s)` part yet. What does it do?

JakeQZ replied:
> 
> 
> With the `=== 1`, we've switched the meaning from "at least one content-type meta tag" to "exactly one content-type meta tag".
> 
> Even if multiple occurrences do not make much sense, I'd still prefer us to recognize this as "there is one" by keeping using `!== false`, and also cover this edge case with a test.

The return value does appear to indicate the number of matches (`0` or `1` because it is limited to one match) with `false` on error.  So to be more explicit we'd probably need something like

```php
$contentTypeMetaTagMatchCount = \preg_match(...);
$hasContentTypeMetaTag = $contentTypeMetaTagMatchCount !== false && $contentTypeMetaTagMatchCount > 0;
```

JakeQZ replied:
> 
> 
> I don't quite understand the `(?=\\s)` part yet. What does it do?

It asserts that at that position the next character is a whitespace but without 'consuming' it ('lookahead').  If it was 'consumed', `\\shttp-equiv` would not match if there was nothing else beetween `meta` and `http-equiv` because the whitespace to be mached there would have already been consumed.  Without asserting `meta` is followed by whitespace, other tag names beginning with `meta` (hitherto as yet not defined) would also be matched.

oliverklee replied:
> So to be more explicit we'd probably need something like
> 
> ```
> $contentTypeMetaTagMatchCount = \preg_match(...);
> $hasContentTypeMetaTag = $contentTypeMetaTagMatchCount !== false && $contentTypeMetaTagMatchCount > 0;
> ```

Or even more explicit:

```php
$hasContentTypeMetaTag = \is_int($contentTypeMetaTagMatchCount) false && $contentTypeMetaTagMatchCount > 0;
```

oliverklee replied:
Thanks for the explanation!

JakeQZ commented:
I've changed the test for the `preg_match` return value as suggested, and also added tests to confirm the regex does not match unexpected longer keywords which contain the expected keyword as a substring.

oliverklee reviewed (approved):
LGTM.
